File: leetcode528/leetcode528_1.cpp
class Solution {
public:
    Solution(vector<int>& w) {
        aggregated_sum = vector<int> (w.size());
        int sum = 0;
        for (int i = 0; i < w.size(); ++i) {
            sum += w[i];
            aggregated_sum[i] = sum;
        }
    }
    
    int pickIndex() {        
        int random = rand() % aggregated_sum.back();
        
        return upper_bound(begin(aggregated_sum), end(aggregated_sum), random) - begin(aggregated_sum);
    }
private:
    vector<int> aggregated_sum;
};

// Runtime: 380 ms, faster than 16.64% of C++ online submissions for Random Pick with Weight.
// Memory Usage: 40.4 MB, less than 55.60% of C++ online submissions for Random Pick with Weight.
// time complexity: Solution O(n) pickIndex O(log n)
// space copmlexity: O(n)

File: leetcode528/leetcode528_1.java
class Solution {
    Random random;
    int[] wSums;
    
    public Solution(int[] w) {
        this.random = new Random();
        for(int i = 1; i < w.length; i++)
            w[i] += w[i - 1];
        this.wSums = w;
    }
    
    public int pickIndex() {
        int nextInt = random.nextInt(wSums[wSums.length - 1]);
        int i = Arrays.binarySearch(wSums, nextInt);
        
        return Math.abs(i + 1);
    }
}

// https://leetcode.com/problems/random-pick-with-weight/discuss/154044/Java-accumulated-freq-sum-and-binary-search
// Runtime: 30 ms, faster than 27.45% of Java online submissions for Random Pick with Weight.
// Memory Usage: 52.8 MB, less than 5.01% of Java online submissions for Random Pick with Weight.
// time complexity: O(log n)
// space complexity: O(n)